CONFIGURATION.md
# Configuration for Dratini

The configuration file format for Dratini is [TOML](https://github.com/toml-lang/toml).

The configuration for Dratini has some sections. The example is [here](test/dratini.toml).

 * [Core Section](#core-section)
 * [iOS Section](#ios-section)
 * [Android Section](#android-section)
 * [Log Section](#log-section)

## Core Section

|name            |type  |description                                 |default         |note                                |
|----------------|------|--------------------------------------------|----------------|------------------------------------|
|workers         |int64 |number of workers for push notification     |runtime.NumCPU()|`-w` options can overwrite          |
|queues          |int64 |size of internal queue for push notification|8192            |`-q` options can overwrite          |
|notification_max|int64 |limit of push notifications once            |100             |                                    |
|pusher_max      |int64 |maximum goroutines for asynchronous pushing |0               |If the value is less than or equal to zero, each worker pushes synchronously|
|shutdown_timeout|int64 |timeout to wait for connections to return to idle when server shutdown (second) | 10 | |

## iOS Section

|name                  |type  |description                                           |default   |note                           |
|----------------------|------|------------------------------------------------------|----------|-------------------------------|
|enabled               |bool  |On/Off for push notication to APNs                    |true      |                               |
|pem_cert_path         |string|certification file path for APNs                      |          |                               |
|pem_key_path          |string|secret key file path for APNs                         |          |                               |
|pem_key_passphrase    |string|secret key file pass phrase for APNs                  |          |                               |
|sandbox               |bool  |On/Off for sandbox environment                        |true      |                               |
|retry_max             |int   |maximum retry count for push notication to APNs       |1         |                               |
|timeout               |int   |timeout for push notification to APNs                 |5         |                               |
|keepalive_timeout     |int   |time for continuing keep-alive connection to APNs     |90        |                               |
|keepalive_conns       |int   |number of keep-alive connection to APNs               |runtime.NumCPU()|                         |
|topic                 |string|the assigned value of `apns-topic` for Request headers|          |                               |

`topic` is mandatory when the client is connected using the certificate that supports multiple topics.

## Android Section

|name             |type  |description                                     |default|note|
|-----------------|------|------------------------------------------------|-------|----|
|enabled          |bool  |On/Off for push notication to GCM               |true   |    |
|apikey           |string|API key string for GCM                          |       |    |
|timeout          |int   |timeout for push notication to GCM              |5(sec) |    |
|keepalive_timeout|int   |time for continuing keep-alive connection to GCM|90     |    |
|keepalive_conns  |int   |number of keep-alive connection to GCM          |runtime.NumCPU()||
|retry_max        |int   |maximum retry count for push notication to GCM  |1      |    |
|use_fcm          |bool  |Use FCM endpoint instead of GCM (by default, `Dratini` uses GCM endpoint)  |false      |    |

## Log Section

|name      |type  |description    |default|note                             |
|----------|------|---------------|-------|---------------------------------|
|access_log|string|access log path|stdout |                                 |
|error_log |string|error log path |stderr |                                 |
|level     |string|log level      |error  |panic,fatal,error,warn,info,debug|

`access_log` and `error_log` are allowed to give not only file-path but `stdout` and `stderr` and `discard`.

pusher/conf.go
package pusher

import (
	"runtime"

	"github.com/BurntSushi/toml"
)

type ConfToml struct {
	Core    SectionCore    `toml:"core"`
	Android SectionAndroid `toml:"android"`
	Ios     SectionIos     `toml:"ios"`
	Log     SectionLog     `toml:"log"`
}

type SectionCore struct {
	WorkerNum       int64 `toml:"workers"`
	QueueNum        int64 `toml:"queues"`
	NotificationMax int64 `toml:"notification_max"`
	PusherMax       int64 `toml:"pusher_max"`
	ShutdownTimeout int64 `toml:"shutdown_timeout"`
}

type SectionAndroid struct {
	Enabled          bool   `toml:"enabled"`
	ApiKey           string `toml:"apikey"`
	Timeout          int    `toml:"timeout"`
	KeepAliveTimeout int    `toml:"keepalive_timeout"`
	KeepAliveConns   int    `toml:"keepalive_conns"`
	RetryMax         int    `toml:"retry_max"`
	UseFCM           bool   `toml:"use_fcm"`
}

type SectionIos struct {
	Enabled          bool   `toml:"enabled"`
	PemCertPath      string `toml:"pem_cert_path"`
	PemKeyPath       string `toml:"pem_key_path"`
	PemKeyPassphrase string `toml:"pem_key_passphrase"`
	Sandbox          bool   `toml:"sandbox"`
	RetryMax         int    `toml:"retry_max"`
	Timeout          int    `toml:"timeout"`
	KeepAliveTimeout int    `toml:"keepalive_timeout"`
	KeepAliveConns   int    `toml:"keepalive_conns"`
	Topic            string `toml:"topic"`
}

type SectionLog struct {
	AccessLog string `toml:"access_log"`
	ErrorLog  string `toml:"error_log"`
	Level     string `toml:"level"`
}

func BuildDefaultConf() ConfToml {
	numCPU := runtime.NumCPU()

	var conf ConfToml
	// Core
	conf.Core.WorkerNum = int64(numCPU)
	conf.Core.QueueNum = 8192
	conf.Core.NotificationMax = 100
	conf.Core.PusherMax = 0
	conf.Core.ShutdownTimeout = 10
	// Android
	conf.Android.ApiKey = ""
	conf.Android.Enabled = true
	conf.Android.Timeout = 5
	conf.Android.KeepAliveTimeout = 90
	conf.Android.KeepAliveConns = numCPU
	conf.Android.RetryMax = 1
	conf.Android.UseFCM = true
	// iOS
	conf.Ios.Enabled = true
	conf.Ios.PemCertPath = ""
	conf.Ios.PemKeyPath = ""
	conf.Ios.Sandbox = true
	conf.Ios.RetryMax = 1
	conf.Ios.Timeout = 5
	conf.Ios.KeepAliveTimeout = 90
	conf.Ios.KeepAliveConns = numCPU
	conf.Ios.Topic = ""
	// log
	conf.Log.AccessLog = "stdout"
	conf.Log.ErrorLog = "stderr"
	conf.Log.Level = "error"
	return conf
}

func LoadConf(confDratini ConfToml, confPath string) (ConfToml, error) {
	_, err := toml.DecodeFile(confPath, &confDratini)
	if err != nil {
		return confDratini, err
	}
	return confDratini, nil
}
